piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
-        } else if point.y >= height {
-            (false, self.line_metrics.last().unwrap())
+        let (is_inside, index, trailing) = self.pango_layout.xy_to_index(x as i32, y as i32);
+        let index = if trailing == 0 {
+            index.try_into().unwrap()
         } else {
-            let line = self
-                .line_metrics
-                .iter()
-                .find(|l| point.y >= l.y_offset && point.y < l.y_offset + l.height)
-                .unwrap();
-            (true, line)
+            /*
+             * NOTE(ForLoveOfCats): The docs specify that a non-zero value for trailing
+             * indicates that the point aligns to the trailing edge of the grapheme. In
+             * that case the value tells us the number of "characters" in the grapheme.
+             */
+
+            let text = self.pango_layout.get_text().unwrap();
+            let index = index.try_into().unwrap();
+            let trailing = trailing.try_into().unwrap();
+
+            text[index..]
+                .char_indices()
+                .nth(trailing)
+                .map(|(offset, _)| index + offset)
+                .unwrap_or_else(|| text.len())
         };
 
-        // Trailing whitespace is remove for the line
-        let line = &self.text[lm.range()];
-
-        let mut htp = hit_test_line_point(&self.font, line, point);
-        htp.idx += lm.start_offset;
-        if htp.idx == lm.end_offset {
-            htp.idx -= util::trailing_nlf(line).unwrap_or(0);
-        }
-        htp.is_inside &= y_inside;
-        htp
-    }
-
-    fn hit_test_text_position(&self, idx: usize) -> HitTestPosition {
-        let idx = idx.min(self.text.len());
-        assert!(self.text.is_char_boundary(idx));
-
-        if idx == 0 && self.text.is_empty() {
-            return HitTestPosition::new(Point::new(0., self.font.extents().ascent), 0);
-        }
-
-        // first need to find line it's on, and get line start offset
-        let line_num = util::line_number_for_position(&self.line_metrics, idx);
-        let lm = self.line_metrics.get(line_num).cloned().unwrap();
-
-        let y_pos = lm.y_offset + lm.baseline;
+        let (metric_index, metric) = self
+            .line_metrics
+            .iter()
+            .enumerate()
+            .find(|(_, metric)| {
+                metric.start_offset <= index
+                    && index <= metric.end_offset - metric.trailing_whitespace
+            })
+            .unwrap();
 
-        // Then for the line, do text position
-        // Trailing whitespace is removed for the line
-        let line = &self.text[lm.range()];
-        let line_position = idx - lm.start_offset;
+        //NOTE: Manually move to start of next line when hit test is at end of line
+        let index = {
+            let text = self.pango_layout.get_text().unwrap();
+            if index == text.len() || matches!(text.as_bytes()[index], b'\r' | b'\n') {
+                index
+            } else {
```

Comment:
Does this match the comment on line 435? I'm a bit confused.

Reply:
This has been simplified and comments improved

Reply:
I've improved the logic around this and improved the comment to be more descriptive of what the goal of that section is